src/Chainweb/Miner/Coordinator.hs

```diff
@@ -232,21 +218,25 @@ chainChoice c choice = case choice of
 publish
     :: LogFunction
     -> CutDb cas
+    -> TVar PrimedWork
     -> MinerId
     -> PayloadData
     -> SolvedWork
     -> IO ()
-publish lf cdb miner pd s = do
+publish lf cdb pwVar miner pd s = do
     c <- _cut cdb
     now <- getCurrentTimeIntegral
     try (extend c pd s) >>= \case
 
         -- Publish CutHashes to CutDb and log success
         Right (bh, Just ch) -> do
+
+            -- reset the primed payload for this cut extension
+            atomically $ modifyTVar pwVar $ resetPrimed miner (_chainId bh)
```

Comment:
This fixes a race in the mining code where a newWork would retrieve an old payload on a new cut. This could happen when a miner would request a primed payload after a new cut arrived but before the primed payload got updated.

There is probably still a race with external cuts, which requires separate attention.